File: leetcode-python/_24xx/_2523_closest_prime_num_in_range.py
import math
import unittest
from typing import List

class Solution:
    def closestPrimes(self, left: int, right: int) -> List[int]:
        def get_primes(left, right):
            nums = [True] * (right + 1)
            nums[0] = False
            nums[1] = False
            for i in range(2, math.floor(math.sqrt(right)) + 1):
                if nums[i] is False:
                    continue
                for j in range(i+i, right + 1, i):
                    nums[j] = False
            ans = []
            for i in range(2, right + 1):
                if nums[i] is True and left <= i <= right:
                    ans.append(i)
            return ans
        nums = get_primes(left, right)
        if len(nums) <= 1:
            return [-1, -1]
        ans = [nums[0], nums[1]]
        d = nums[1] - nums[0]
        for i in range(len(nums) -1):
            if nums[i+1] - nums[i] < d:
                d = nums[i+1] - nums[i]
                ans = [nums[i], nums[i+1]]
        return ans

class MyTestCase(unittest.TestCase):

    def setUp(self) -> None:
        self.s = Solution()
    
    def test_case_1(self):
        actual = self.s.closestPrimes(left = 10, right = 19)
        self.assertEqual([11, 13], actual)
        
    def test_case_2(self):
        actual = self.s.closestPrimes(left = 4, right = 6)
        self.assertEqual([-1, -1], actual)

    def test_case_3(self):
        actual = self.s.closestPrimes(left = 2, right = 11)
        self.assertEqual([2, 3], actual)

    def test_case_3(self):
        actual = self.s.closestPrimes(left = 1, right = 1000000)
        self.assertEqual([2, 3], actual)

    def test_case_4(self):
        actual = self.s.closestPrimes(left = 1, right = 1)
        self.assertEqual([-1, -1], actual)

if __name__ == '__main__':
    unittest.main()

File: leetcode-cpp/_24xx/_2523_closest_prime_num_in_range.cpp
#include "../test/catch_amalgamated.hpp"

using namespace std;

class Solution
{
public:
    vector<int> closestPrimes(int left, int right)
    {
        auto primes = getPrimes(left, right);
        if (primes.size() <= 1)
        {
            return vector<int>{-1, -1};
        }
        auto ans = vector<int>{primes.at(0), primes.at(1)};
        int diff = primes.at(1) - primes.at(0);
        for (int i = 0; i < primes.size() - 1; i++)
        {
            if (primes.at(i + 1) - primes.at(i) < diff)
            {
                ans.at(0) = primes.at(i);
                ans.at(1) = primes.at(i + 1);
                diff = primes.at(i + 1) - primes.at(i);
            }
        }
        return ans;
    }

    vector<int> getPrimes(int left, int right)
    {
        vector<bool> nums(right + 1, true);
        nums.at(0) = false;
        nums.at(1) = false;
        for (int i = 2; i < (int)sqrt(right) + 1; i++)
        {
            if (nums.at(i) == false)
            {
                continue;
            }
            for (int j = i + i; j < right + 1; j = j + i)
            {
                nums.at(j) = false;
            }
        }

        vector<int> primes{};
        for (size_t i = 0; i < right + 1; i++)
        {
            if (nums.at(i) == true && i >= left && i <= right)
            {
                primes.push_back(i);
            }
        }
        return primes;
    }
};

TEST_CASE("test_case_1")
{
    Solution s;
    auto actual = s.closestPrimes(10, 19);
    auto expect = vector<int>{11, 13};
    CHECK(actual == expect);
}

TEST_CASE("test_case_2")
{
    Solution s;
    auto actual = s.closestPrimes(4, 6);
    auto expect = vector<int>{-1, -1};
    CHECK(actual == expect);
}

TEST_CASE("test_case_3")
{
    Solution s;
    auto actual = s.closestPrimes(1, 1);
    auto expect = vector<int>{-1, -1};
    CHECK(actual == expect);
}

TEST_CASE("test_case_4")
{
    Solution s;
    auto actual = s.closestPrimes(1, 1000000);
    auto expect = vector<int>{2, 3};
    CHECK(actual == expect);
}
